doc/chmconvert.1
.\" -*- nroff -*-
.\"
.\"  chmconvert.1 - Manual page for chmconvert
.\"
.\"  Copyright (c) 2006-2012 Sebastien Maret
.\" 
.\"  This file is part of Astrochem.
.\"
.\"  Astrochem is free software: you can redistribute it and/or modify
.\"  it under the terms of the GNU General Public License as published
.\"  by the Free Software Foundation, either version 3 of the License,
.\"  or (at your option) any later version.
.\"
.\"  Astrochem is distributed in the hope that it will be useful, but
.\"  WITHOUT ANY WARRANTY; without even the implied warranty of
.\"  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\"  General Public License for more details.
.\" 
.\"  You should have received a copy of the GNU General Public License
.\"  along with Astrochem.  If not, see <http://www.gnu.org/licenses/>.
.\"
.Dd June 3, 2011
.Os
.Dt CHMCONVERT 1
.Sh NAME
.Nm chmconvert
.Nd convert chemical networks into .chm format
.Sh SYNOPSIS
.Nm
.Op Ar option
.Op Ar file
.\"
.\" Description
.\"
.Sh DESCRIPTION
.Nm
is a tool to convert chemical network files from different formats to
.Xr astrochem 1 
native format
.Em .chm .
The format of the input file is determined
from the extension of the input file. Supported formats are
.Em .osu
(Ohio State University astrochemistry database) and
.Em .kida
(KInetic Database for Astrochemistry).
.\"
.\" Options
.\"
.Sh OPTIONS
.Bl -tag -width flag
.It Cm -h, --help
Display help text.
.It Cm -V, --version
Display astrochem version information.
.It Cm -o Ar file , Cm --output Ar file
Write the converted network in
.Ar file "."
Default is the standard output.
.El
.\"
.\" Authors, copyright, and see also
.\"
.Sh AUTHOR
Written by Sebastien Maret.
.Sh "REPORTING BUGS"
Report bugs to
.Aq sebastien.maret@obs.ujf-grenoble.fr .
.Sh COPYRIGHT
Copyright \(co 2006-2012 Sebastien Maret
.Pp
This is free software. You may redistribute copies of it under the
terms of the GNU General Public License. There is NO WARRANTY, to the
extent permitted by law.
.Sh "SEE ALSO"
.Xr astrochem 1 ","
.Xr plabun 1 ","
.Xr plroute 1


NEWS
What's new in Astrochem ?
=========================

astrochem-0.4 (June 3, 2011)
----------------------------

* The main destruction/formation channels of a species plotted by
  Plroute are now ordered by their total contribution to the
  formation/destruction of that species over the time of the
  computation. An aption has been added to print the relative
  contribution of each channel.

* Fixed a bug in the computation of the main destruction routes.

astrochem-0.3 (February 28, 2011)
---------------------------------

* The rate for cosmic-ray desorption is now either 1) computed
  following Hasegawa & Herbst (1993) or 2) given explicitly in the
  network file (e.g. using values from Bringa & Johnson 2004).

* If the LAPACK library is installed, Astrochem can now use the
  CVLapackDense solver from SUNDIALS instead of CVDense. This results
  in a performance increase of 30% or more, depending on the LAPACK
  version.

astrochem-0.2 (January 21, 2011)
--------------------------------

* First public release


src/chmconvert.py
#!@PYTHON@
#
#  chmconvert - Convert chemical networks into .chm format
#
#  Copyright (c) 2006-2011 Sebastien Maret
# 
#  This file is part of Astrochem.
#
#  Astrochem is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published
#  by the Free Software Foundation, either version 3 of the License,
#  or (at your option) any later version.
#
#  Astrochem is distributed in the hope that it will be useful, but
#  WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
#  General Public License for more details.
# 
#  You should have received a copy of the GNU General Public License
#  along with Astrochem.  If not, see <http://www.gnu.org/licenses/>.

import sys
import getopt
from libastrochem import network

def usage():
    """
    Display usage.

    """

    print """Usage: chmconvert [option] [file]

Common options:
   -h, --help         Display this help
   -V, --version      Display chmconvert version information
   -o, --output       Write the edited network in a file

See chmconvert(1) man page for a complete list of commands and options.
Report bugs to <sebastien.maret@obs.ujf-grenoble.fr>."""

def version():
    """
    Display version number.

    """

    print "This is chmconvert, version %s" % VERSION
    print """Copyright (c) 2006-2011 Sebastien Maret

This is free software. You may redistribute copies of it under the terms
of the GNU General Public License. There is NO WARRANTY, to the extent
permitted by law."""

def main():
    """
    Main function for chmconvert

    """
    
    # Parse options and check arguments. Display help message if
    # an unknown option is given.
    try:
	opts, args = getopt.getopt(sys.argv[1:], "hVo:",
				   ["help", "version", "output"])
    except getopt.GetoptError:
	usage()
	sys.exit(1)

    output = None
    for opt, arg in opts:
	if opt in ("-h", "--help"):
	    usage()
	    sys.exit()
	if opt in ("-V", "--version"):
	    version()
	    sys.exit()
	if opt in ("-o", "--output"):
	    output = arg

    # Check that we have a least a filename
    if len(args) == 1:
        network_file = args[0]
    else:
        sys.stderr.write("chmconvert: no file to convert.\n")
        sys.stderr.write("Type \'chmconvert --help\' for more information.\n")
        sys.exit(1)
        
    # Guess the format from the file extension
    if len(network_file.rsplit('.', 1)) == 2:
        network_file_base, network_file_ext = network_file.rsplit('.', 1)
    else:
        sys.stderr.write("chmconvert: file has no extension.\n")
        sys.exit(1)  
    if network_file_ext in ["osu", "kida"]:
        format = network_file_ext
    else:
        sys.stderr.write("chmconvert: unknown network format \"%s\".\n" 
                         % network_file_ext)
        sys.exit(1)
		
    # Open input and output files. We set fileout to stdout if
    # the --output option was not set.
    try:
        filein = open(network_file)
    except:
        sys.stderr.write("chmconvert: can't open %s.\n" % network_file)
        sys.exit(1)
    if not output:
        fileout = sys.stdout
    else:		
        try:
            fileout = open(output, 'w')
        except:
            sys.stderr.write("chmconvert: can't open %s.\n" % output)
            sys.exit(1)
	    
    # Read the network file and convert it
    try:
        net = network(filein, format = format)
    except Exception, err:
        sys.stderr.write("chmconvert: %s.\n" % err)
        sys.exit(1)        
    filein.close()

    # Write the converted file:
    net.write(fileout)
    fileout.close()
    
if __name__ == "__main__":
    main()
